# Table of Contents

[Group 1: Basic Statistics and Demographics](##Group-1:-Basic-Statistics-and-Demographics)<br>
&nbsp;&nbsp;[1) What are the Unique Combinations of Custody Level and Race Among the Inmates?](###What-are-the-Unique-Combinations-of-Custody-Level-and-Race-Among-the-Inmates)<br>
&nbsp;&nbsp;[2) How many Inmates of Each Age Group (e.g., 18-25, 26-35, etc.) are in Custody?](###How-many-Inmates-of-Each-Age-Group-(e.g.,-18-25,-26-35,-etc.)-are-in-Custody)<br>
&nbsp;&nbsp;[3) How many Inmates were Admitted in Each Month of the Year?](###How-many-Inmates-were-Admitted-in-Each-Month-of-the-Year?)<br>
&nbsp;&nbsp;[4) How many Inmates were Admitted on Each Day of the Week?](###How-many-Inmates-were-Admitted-on-Each-Day-of-the-Week?)<br>
&nbsp;&nbsp;[5) How many Inmates are there for Each Unique Combination of Gender and Race?](###How-many-Inmates-are-there-for-Each-Unique-Combination-of-Gender-and-Race)<br>
&nbsp;&nbsp;[6) How many Inmates are there for Each Unique Combination of Custody Level and Gender?](###How-many-Inmates-are-there-for-Each-Unique-Combination-of-Custody-Level-and-Gender?)<br>
[Group 2: Average Age Analysis](##Group-2:-Average-Age-Analysis)<br>
&nbsp;&nbsp;[7) What is the Average Age of Inmates for Each Race?](###What-is-the-Average-Age-of-Inmates-for-Each-Race?)<br>
&nbsp;&nbsp;[8) What is the Average Age of Inmates for Each Gender?](###What-is-the-Average-Age-of-Inmates-for-Each-Gender?)<br>
&nbsp;&nbsp;[9) What is the Average Age of Inmates for Each Custody Level?](###What-is-the-Average-Age-of-Inmates-for-Each-Custody-Level?)<br>
&nbsp;&nbsp;[10) What is the Average Age of Inmates with Gang Affiliation Compared to No Gang Affiliation?](###What-is-the-Average-Age-of-Inmates-with-Gang-Affiliation-Compared-to-No-Gang-Affiliation?)<br>
&nbsp;&nbsp;[11) Average Age by Custody Level and Gang Affiliation](###Average-Age-by-Custody-Level-and-Gang-Affiliation)<br>
&nbsp;&nbsp;[12) Average Age by Infraction Occurrence](###Average-Age-by-Infraction-Occurrence)<br>
&nbsp;&nbsp;[13) Average Age by Mental Health](###Average-Age-by-Mental-Health)<br>
[Group 3: Top Charge Analysis](##Group-3:-Top-Charge-Analysis)<br>
&nbsp;&nbsp;[14) What is the Most Common Top Charge Among Inmates by Gender?](###What-is-the-Most-Common-Top-Charge-Among-Inmates-by-Gender?)<br>
&nbsp;&nbsp;[15) Top Charge by Race](###Top-Charge-by-Race)<br>
&nbsp;&nbsp;[16) Top Charge by Age](###Top-Charge-by-Age)<br>
&nbsp;&nbsp;[17) Top Charge by Custody Level](###Top-Charge-by-Custody-Level)<br>
&nbsp;&nbsp;[18) Top Charge and Gang Affiliation](###Top-Charge-and-Gang-Affiliation)<br>
&nbsp;&nbsp;[19) Top Charge and Mental Health Status](###Top-Charge-and-Mental-Health-Status)<br>
[Group 4: Infraction Analysis](##Group-4:-Infraction-Analysis)<br>
&nbsp;&nbsp;[20) Infraction Occurrence for Inmates of More Than 5 Years](###Infraction-Occurrence-for-Inmates-of-More-Than-5-Years)<br>
&nbsp;&nbsp;[21) Infraction Occurrence for Inmates of Less Than 30 Days](###Infraction-Occurrence-for-Inmates-of-Less-Than-30-Days)<br>
&nbsp;&nbsp;[22) Infraction by Mental Health](###Infraction-by-Mental-Health)<br>
&nbsp;&nbsp;[23) Infraction by Gang Affiliation](###Infraction-by-Gang-Affiliation)<br>
[Group 5: Inmate Custody Duration](##Group-5:-Inmate-Custody-Duration)<br>
&nbsp;&nbsp;[24) How Many Inmates have been in Custody for More Than a Year](###How-Many-Inmates-have-been-in-Custody-for-More-Than-a-Year?)<br>
&nbsp;&nbsp;[24) How Many Inmates have been in Custody for More Than a Month](###How-Many-Inmates-have-been-in-Custody-for-More-Than-a-Month?)<br>
&nbsp;&nbsp;[24) How Many Inmates have been in Custody for More Than 3 Months](###How-Many-Inmates-have-been-in-Custody-for-More-Than-3-Months?)<br>
&nbsp;&nbsp;[24) How Many Inmates have been in Custody for More Than 6 Months](###How-Many-Inmates-have-been-in-Custody-for-More-Than-6-Months?)<br>
&nbsp;&nbsp;[24) How Many Inmates have been in Custody for More Than 9 Months](###How-Many-Inmates-have-been-in-Custody-for-More-Than-9-Months?)<br>
[Group 6: Miscellaneous](##Group-6:-Miscellaneous)<br>
&nbsp;&nbsp;[25) How Many Inmates Under Mental Observation are there for Each Custody Level?](###How-Many-Inmates-Under-Mental-Observation-are-there-for-Each-Custody-Level?)<br>
&nbsp;&nbsp;[26) What are the Top 10 Charges for the Inmates Under Mental Health Observation?](###What-are-the-Top-10-Charges-for-the-Inmates-Under-Mental-Health-Observation?)<br>
&nbsp;&nbsp;[27) Gang Affiliations by Gender](###Gang-Affiliations-by-Gender)<br>
&nbsp;&nbsp;[28) Inmates Under Mental Health Observation by Race](###Inmates-Under-Mental-Health-Observation-by-Race) <br>
&nbsp;&nbsp;[29) Mental Health Status by Gang Affiliation](###Mental-Health-Status-by-Gang-Affiliation)<br>


If the hyperlinks are not working correctly, reload the file and try again.


**Importing** the Necessary Libraries

In [1]:
import pandas as pd
import sqlite3

**Reading In** the CSV that was created in the previous file

In [9]:
df = pd.read_csv('criminals.csv')
df.head()

CUSTODY_LEVEL BRADH RACE GENDER  AGE INMATE_STATUS_CODE SRG_FLG  TOP_CHARGE  \
0           MAX     Y    B      M   21                 DE       Y  110-125.25   
1           MED     Y    O      M   70                 DE       N      130.50   
2           MED     Y    O      M   31                 DE       Y      220.16   
3           NaN     Y    O      M   23                 DE       N  110-125.25   
4           MED     N    O      M   32                 DE       N  110-125.25   

  INFRACTION        Date      Time  
0          Y  2023-03-23  14:49:00  
1          N  2023-03-16  21:14:00  
2          N  2023-06-11  00:38:00  
3          Y  2023-03-02  01:57:34  
4          N  2021-09-05  00:50:57

### Establish a connection to the SQLite database file 'crime_database.db'

You can create your own database file by running the code below and it'll run properly. The naming of your *.db* file can be whatever you want

In [3]:
conn = sqlite3.connect('crime_database.db')

### Convert the DF to a SQL Table

In [4]:
df.to_sql('mytable', conn, if_exists='replace', index=False)

6146

## Group 1: Basic Statistics and Demographics

### What are the Unique Combinations of Custody Level and Race Among the Inmates?

In [21]:
query = """SELECT DISTINCT CUSTODY_LEVEL, RACE, COUNT(*) as Number_of_Prisoners,
            COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
            FROM mytable
            WHERE CUSTODY_LEVEL != 'None' AND RACE != 'None'
            GROUP BY CUSTODY_LEVEL, RACE
            ORDER BY COUNT(*) DESC;"""
result = pd.read_sql_query(query, conn)

print(result)

   CUSTODY_LEVEL RACE  Number_of_Prisoners  Percentage
0            MED    B                 1510   24.568825
1            MAX    B                 1226   19.947934
2            MIN    B                  923   15.017898
3            MED    O                  793   12.902701
4            MIN    O                  535    8.704849
5            MAX    O                  427    6.947608
6            MED    W                  229    3.726001
7            MIN    W                  191    3.107712
8            MAX    W                   89    1.448096
9            MED    A                   66    1.073869
10           MIN    A                   58    0.943703
11           MAX    A                   21    0.341686
12           MED    I                    9    0.146437
13           MIN    I                    7    0.113895
14           MAX    I                    3    0.048812
15           MAX    U                    3    0.048812
16           MIN    U                    3    0.048812
17        

The number of prisoners by custody level and race just ends up being ordered by their race. For all races except the **Unknown** category, most people were incarcerated in **Medium** level security prisons.

### How many Inmates of Each Age Group (e.g., 18-25, 26-35, etc.) are in Custody?

In [23]:
query = """SELECT CASE
                WHEN AGE BETWEEN 18 and 25 THEN '18-25'
                WHEN AGE BETWEEN 26 and 35 THEN '26-35'
                WHEN AGE BETWEEN 36 and 45 THEN '36-45'
                WHEN AGE BETWEEN 46 and 55 THEN '46-55'
                WHEN AGE BETWEEN 56 and 65 THEN '56-65'
                ELSE '66 and above'
                END as Age_Group,
                COUNT(*) AS Number_of_Inmates,
                COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
                FROM mytable
                GROUP BY Age_GROUP;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

      Age_Group  Number_of_Inmates  Percentage
0         18-25               1329   21.623820
1         26-35               2140   34.819395
2         36-45               1456   23.690205
3         46-55                738   12.007810
4         56-65                408    6.638464
5  66 and above                 75    1.220306


The **26-35** age range has the most number of inmates. Roughly and equal amount of inmates are in the surrounding age ranges of 18-25 and 36-45. 

### How many Inmates were Admitted in Each Month of the Year?

In [25]:
query ="""SELECT strftime('%m', Date)  as Month_Number, 
           COUNT(*) AS Charge_Count,
           COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
           FROM mytable
           GROUP BY Month_Number
           ORDER BY COUNT(*) DESC;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Month_Number  Charge_Count  Percentage
0            06           987   16.059226
1            07           981   15.961601
2            05           811   13.195574
3            04           613    9.973967
4            03           554    9.013993
5            01           436    7.094045
6            02           396    6.443215
7            12           324    5.271721
8            11           284    4.620892
9            10           271    4.409372
10           09           254    4.132769
11           08           235    3.823625


The late-beginning and middle of the year have the most amount of admitted inmates. It's surprising to see that June and July have the highest number of inmates when August and September then have the lowest number of inmates. There's almost a 4x difference between the number of inmates in June and July compared to August and September. 

SQLite does not have built in functionality for the *MONTH()* function. I will show the normal SQL code I would have used below:

In [ ]:
SELECT COUNT(*), MONTH(Date) as Month_Number 
    FROM mytable
    GROUP BY Month_Number;

### How many Inmates were Admitted on Each Day of the Week?

In [26]:
query = """SELECT CASE strftime('%w', Date) 
            WHEN '0' THEN 'Sunday'
            WHEN '1' THEN 'Monday'
            WHEN '2' THEN 'Tuesday'
            WHEN '3' THEN 'Wednesday'
            WHEN '4' THEN 'Thursday'
            WHEN '5' THEN 'Friday'
            WHEN '6' THEN 'Saturday'
            END AS Day_of_Week_Number, 
            COUNT(*) AS Number_of_Inmates,
            COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
            FROM mytable
            GROUP BY Day_of_Week_Number
            ORDER BY COUNT(*) DESC;
            """
        
result = pd.read_sql_query(query, conn)
print(result)

  Day_of_Week_Number  Number_of_Inmates  Percentage
0           Thursday               1176   19.134396
1          Wednesday               1096   17.832737
2             Friday               1060   17.246990
3            Tuesday                876   14.253173
4           Saturday                693   11.275626
5             Monday                664   10.803775
6             Sunday                581    9.453303


Most of the inmates were brought in during the latter half of the workweek. I find it intriguing how the weekend has much lower numbers of arrests.

### How many Inmates are there for Each Unique Combination of Gender and Race?

In [28]:
query = """SELECT DISTINCT GENDER, RACE, COUNT(*) as Number_of_Prisoners,
            COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
            FROM mytable
            WHERE GENDER != 'None' AND RACE != 'None'
            GROUP BY GENDER, RACE
            ORDER BY COUNT(*) DESC;"""
result = pd.read_sql_query(query, conn)

print(result)

   GENDER RACE  Number_of_Prisoners  Percentage
0       M    B                 3458   56.264237
1       M    O                 1679   27.318581
2       M    W                  455    7.403189
3       F    B                  220    3.579564
4       M    A                  130    2.115197
5       F    O                   84    1.366743
6       F    W                   58    0.943703
7       M    I                   17    0.276603
8       F    A                   16    0.260332
9       M    U                    7    0.113895
10      F    I                    2    0.032541
11      F    U                    2    0.032541


These results are not that surprising when you consider how the majority of the inmates in this dataset are Male and Black. Still though, over 56% of the dataset being black males is extremely high. For every race, the Males exceed the number of Females. 

### How many Inmates are there for Each Unique Combination of Custody Level and Gender?

In [30]:
query = """SELECT DISTINCT GENDER, CUSTODY_LEVEL, COUNT(*) as Number_of_Prisoners,
            COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
            FROM mytable
            WHERE GENDER != 'None' AND CUSTODY_LEVEL != 'None'
            GROUP BY GENDER, CUSTODY_LEVEL
            ORDER BY COUNT(*) DESC;"""
result = pd.read_sql_query(query, conn)

print(result)

  GENDER CUSTODY_LEVEL  Number_of_Prisoners  Percentage
0      M           MED                 2461   40.042304
1      M           MAX                 1651   26.863000
2      M           MIN                 1612   26.228441
3      F           MED                  148    2.408070
4      F           MAX                  118    1.919948
5      F           MIN                  105    1.708428


Both Genders follow the same pattern. One thing of note is that as a percentage, more women than men are in maximum security.

## Group 2: Average Age Analysis

### What is the Average Age of Inmates for Each Race?

In [32]:
query = """SELECT ROUND(AVG(AGE),2) AS Average_Age, RACE
            FROM mytable
            WHERE RACE != 'None'
            GROUP BY RACE
            ORDER BY Average_Age DESC;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Average_Age RACE
0        39.36    W
1        36.42    A
2        35.95    I
3        35.55    O
4        35.27    B
5        33.78    U


Among the known races, White inmates are the oldest on average by about 3 years. Besides the 8 inmates who fall under the **Unknown** category, the Black inmates have the lowest average age.

### What is the Average Age of Inmates for Each Gender?

In [76]:
query = """SELECT ROUND(AVG(AGE),2) AS Average_Age, GENDER
            FROM mytable
            WHERE GENDER != 'None'
            GROUP BY GENDER
            ORDER BY Average_Age DESC;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Average_Age GENDER
0        35.73      F
1        35.72      M


The average age for men and women in this dataset is virtually the same thing

### What is the Average Age of Inmates for Each Custody Level?

In [8]:
query = """SELECT ROUND(AVG(AGE), 2) as Average_Age, CUSTODY_LEVEL
            FROM mytable
            WHERE CUSTODY_LEVEL != 'None'
            GROUP BY CUSTODY_LEVEL;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Average_Age CUSTODY_LEVEL
0        30.38           MAX
1        35.59           MED
2        41.42           MIN


The average ages of the inmates decreases as the security level increases. Specifically, the minimum security facilities have an average age of 11 years older than maximum security facilities.

###  What is the Average Age of Inmates with Gang Affiliation Compared to No Gang Affiliation?

In [5]:
query = """SELECT ROUND(AVG(AGE), 2) as Average_Age, SRG_FLG
            FROM mytable
            GROUP BY SRG_FLG;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Average_Age SRG_FLG
0        37.72       N
1        28.02       Y


Inmates that have a certified gang affiliation are nearly 10 years younger than non-gang members.

### Average Age by Custody Level and Gang Affiliation

In [10]:
query = """SELECT CUSTODY_LEVEL, SRG_FLG, COUNT(*) AS COUNT, ROUND(AVG(AGE), 2) as Average_Age
            FROM mytable
            WHERE CUSTODY_LEVEL != 'None'
            GROUP BY SRG_FLG, CUSTODY_LEVEL
            ORDER BY Average_Age;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  CUSTODY_LEVEL SRG_FLG  COUNT  Average_Age
0           MAX       Y    853        27.81
1           MED       Y    353        28.24
2           MIN       Y     52        29.75
3           MAX       N    920        32.76
4           MED       N   2261        36.73
5           MIN       N   1674        41.78


For both the gang affiliated inmates and the non-gang affiliated inmates, the youngest inmates are in the maximum security facilities. Another interesting insight from this query is that among gang members, a majority are in maximum security prisons. However, among the non-affiliated inmates, a minority of the inmates are in maximum security institutions. The age distribution for gang affiliated inmates is also more tightly compacted than for the non-affiliated inmates.

### Average Age by Infraction Occurrence

In [11]:
query = """SELECT ROUND(AVG(AGE), 2) AS AVERAGE_AGE, INFRACTION
            FROM mytable
            GROUP BY INFRACTION
            ;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   AVERAGE_AGE INFRACTION
0        37.74          N
1        31.34          Y


On average, the prisoners who committed an infraction are 6 years younger than those who haven't carried out an infraction.

### Average Age by Mental Health 

In [12]:
query = """SELECT ROUND(AVG(AGE), 2) AS Average_Age, BRADH
            FROM mytable
            WHERE BRADH IS NOT NULL
            GROUP BY BRADH
        """

result = pd.read_sql_query(query, conn)
print(result)

   Average_Age BRADH
0        35.53     N
1        35.91     Y


There's not much of a difference between the average ages of inmates who are under mental health observations and those who aren't.

## Group 3: Top Charge Analysis 

### What is the Most Common Top Charge Among Inmates by Gender?

In [80]:
query = """SELECT GENDER, TOP_CHARGE, Charge_Count
            FROM (
                SELECT GENDER, TOP_CHARGE, COUNT(*) AS Charge_Count, 
                ROW_NUMBER() OVER (PARTITION BY GENDER ORDER BY COUNT(*) DESC) AS Rank
                FROM mytable
                WHERE GENDER IS NOT NULL
                GROUP BY GENDER, TOP_CHARGE
            ) 
            WHERE Rank = 1
            ORDER BY Charge_Count DESC, GENDER
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  GENDER TOP_CHARGE  Charge_Count
0      M     125.25           839
1      F     125.25            51


Here's a breakdown of the definition of each top charge according to [NYSenate.gov](https://www.nysenate.gov/legislation/laws/search/):
- 125.25 - Second Degree Murder

Second Degree Murder is the most common charge among both genders.

### Top Charge by Race

In [61]:
query = """SELECT RACE, TOP_CHARGE, Charge_Count
            FROM (
                SELECT RACE, TOP_CHARGE, COUNT(*) AS Charge_Count, 
                ROW_NUMBER() OVER (PARTITION BY RACE ORDER BY COUNT(*) DESC) AS Rank
                FROM mytable
                WHERE RACE IS NOT NULL
                GROUP BY RACE, TOP_CHARGE
            ) 
            WHERE Rank = 1
            ORDER BY Charge_Count DESC, RACE 
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  RACE TOP_CHARGE  Charge_Count
0    B     125.25           580
1    O     125.25           235
2    W     125.25            50
3    A     125.25            23
4    I     215.51             2
5    U     160.15             2


Here's the breakdown of the above [charges](https://www.nysenate.gov/legislation/laws/search/):
* 125.25 - Second Degree Murder
* 215.51 - First Degree Criminal Contempt
* 160.15 - First Degree Robbert

2nd Degree Murder is the most prominent crime for 4 out of 6 of the race categories.

### Top Charge by Age 

In [37]:
query = """SELECT Age_Group, TOP_CHARGE, Charge_Count
FROM (
    SELECT Age_Group, TOP_CHARGE, Charge_Count,
           ROW_NUMBER() OVER (PARTITION BY Age_Group ORDER BY Charge_Count DESC) AS Rank
    FROM (
        SELECT CASE
                   WHEN AGE BETWEEN 18 and 25 THEN '18-25'
                   WHEN AGE BETWEEN 26 and 35 THEN '26-35'
                   WHEN AGE BETWEEN 36 and 45 THEN '36-45'
                   WHEN AGE BETWEEN 46 and 55 THEN '46-55'
                   WHEN AGE BETWEEN 56 and 65 THEN '56-65'
                   ELSE '66 and above'
               END as Age_Group,
               TOP_CHARGE,
               COUNT(*) AS Charge_Count
        FROM mytable
        GROUP BY Age_Group, TOP_CHARGE
    )
)
WHERE Rank = 1
ORDER BY Age_Group, Charge_Count DESC;

"""

result = pd.read_sql_query(query, conn)
print(result)


      Age_Group TOP_CHARGE  Number_of_Inmates
0         18-25     125.25                251
1         26-35     125.25                340
2         36-45     125.25                166
3         46-55     125.25                 86
4         56-65     140.20                 34
5  66 and above     125.25                 17


Here's a breakdown of the [charges](https://www.nysenate.gov/legislation/laws/search/):

* 125.25 - Second Degree Murder
* 140.20 - Third Degree Burglary

Again, Second Degree Murder is the most prominent crime among the different groups. 

### Top Charge by Custody Level

In [63]:
query = """SELECT CUSTODY_LEVEL, TOP_CHARGE, Charge_Count
            FROM (
                SELECT CUSTODY_LEVEL, TOP_CHARGE, COUNT(*) AS Charge_Count,
                    RANK() OVER (PARTITION BY CUSTODY_LEVEL ORDER BY COUNT(*) DESC) AS Rank
                FROM mytable 
                WHERE CUSTODY_LEVEL IS NOT NULL
                GROUP BY CUSTODY_LEVEL, TOP_CHARGE
            ) ranked
            WHERE Rank = 1
            ORDER BY CUSTODY_LEVEL, Charge_Count DESC;
            """

result = pd.read_sql_query(query, conn)
print(result)

  CUSTODY_LEVEL TOP_CHARGE  Charge_Count
0           MAX     125.25           383
1           MED     125.25           503
2           MIN     120.05           155


Here's a breakdown of the [charges](https://www.nysenate.gov/legislation/laws/search/):
* 125.25 - Second Degree Murder
* 120.05 - Second Degree Assault

Again, Second Degree Murder is the most prominent crime among the different groups. 

### Top Charge and Gang Affiliation

In [66]:
query = """SELECT SRG_FLG, TOP_CHARGE, Charge_Count
            FROM (
                SELECT SRG_FLG, TOP_CHARGE, COUNT(*) AS Charge_Count,
                    RANK() OVER (PARTITION BY SRG_FLG ORDER BY COUNT(*) DESC) AS Rank
                FROM mytable 
                GROUP BY SRG_FLG, TOP_CHARGE
            ) ranked
            WHERE Rank = 1
            ORDER BY SRG_FLG, Charge_Count DESC;
            """

result = pd.read_sql_query(query, conn)
print(result)

  SRG_FLG TOP_CHARGE  Charge_Count
0       N     125.25           591
1       Y     125.25           299


The top charge for both gang affiliated and non-gang affiliated inmates is Second Degree Murder

### Top Charge and Mental Health Status 

In [67]:
query = """SELECT SRG_FLG, TOP_CHARGE, Charge_Count
            FROM (
                SELECT SRG_FLG, TOP_CHARGE, COUNT(*) AS Charge_Count,
                    RANK() OVER (PARTITION BY SRG_FLG ORDER BY COUNT(*) DESC) AS Rank
                FROM mytable 
                GROUP BY SRG_FLG, TOP_CHARGE
            ) ranked
            WHERE Rank = 1
            ORDER BY SRG_FLG, Charge_Count DESC;
            """

result = pd.read_sql_query(query, conn)
print(result)

  BRADH TOP_CHARGE  Charge_Count
0     N     125.25           351
1     Y     125.25           539


Once, again, Second Degree Murder is the top charge for inmates under mental health observation and those who aren't.

## Group 4: Infraction Analysis

### Infraction Occurrence for Inmates of More Than 5 Years

In [7]:
query = """SELECT DISTINCT INFRACTION, COUNT(*) AS Inmates_More_Than_5_Years
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 1825
            GROUP BY INFRACTION
            ;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

query = """SELECT COUNT(*) as Number_of_Prisoners 
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 1825
            GROUP BY INFRACTION
            ;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  INFRACTION  Inmates_More_Than_5_Years
0          Y                         28
   Number_of_Prisoners
0                   28


Every inmate of longer than 5 years has committed and infraction. The second query is there to confirm that the 28 queried inmates from above was the entire inmates population of > 5 years.

### Infraction Occurrence for Inmates of Less Than 30 Days

In [5]:
query = """SELECT DISTINCT INFRACTION, COUNT(*)
            FROM mytable
            WHERE julianday('now') - julianday(Date) < 30
            GROUP BY INFRACTION
            ;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  INFRACTION  COUNT(*)
0          N       661


Zero of the inmates who were admitted in the past 30 days have committed an infraction. It's important to note that the code above is dynamic in the sense that the *COUNT(*) will change based on the day you run the code. The number above is as of August 3rd, 2023.

### Infraction by Mental Health

In [7]:
query = """SELECT INFRACTION, BRADH, COUNT(*) AS Charge_Count
            FROM mytable
            GROUP BY INFRACTION, BRADH
            ORDER BY Charge_Count DESC;
        """

result = pd.read_sql_query(query, conn) 
print(result)

infraction_mental_obs = (1303/(1303+626)) * 100
non_infraction_mental_obs = (1930/(1930+2287)) * 100

print('\n')
print('No Infraction and Mental Health Percentage',non_infraction_mental_obs)
print('Infraction and Mental Health Percentage',infraction_mental_obs)

  INFRACTION BRADH  Charge_Count
0          N     N          2287
1          N     Y          1930
2          Y     Y          1303
3          Y     N           626


No Infraction and Mental Health Percentage 45.76713303296182
Infraction and Mental Health Percentage 67.54795230689477


About 68% of inmates who committed an infraction are under mental health observations compared to 46% of inmates who didn't commit any infractions. It's possible that part of the repercussions of committing an infraction could be to be placed under mental health observation. We have no way of knowing what the exact infraction was, so it may be perfectly suitable for an inmate to need extra observation depending on the infraction. 

### Infraction by Gang Affiliation

In [8]:
query = """SELECT INFRACTION, SRG_FLG, COUNT(*) AS Charge_Count
            FROM mytable
            GROUP BY INFRACTION, SRG_FLG
            ORDER BY Charge_Count DESC;
        """

result = pd.read_sql_query(query, conn) 
print(result)

infraction_gang = (791/(791+469)) * 100
infraction_non_gang = (1138/(1138+3748)) * 100

print('\n')
print('Infraction and Gang Affiliated',infraction_gang)
print('Infraction and Non-Gang Affiliated',infraction_non_gang)

  INFRACTION SRG_FLG  Charge_Count
0          N       N          3748
1          Y       N          1138
2          Y       Y           791
3          N       Y           469


Infraction and Gang Affiliated 62.77777777777778
Infraction and Non-Gang Affiliated 23.29103561195252


There is a large difference in the propensity of gang affiliated members to commit an infraction compared to non-gang affiliated inmates. 

## Group 5: Inmate Custody Duration

### How Many Inmates have been in Custody for More Than a Year?

In [38]:
query = """SELECT COUNT(*) AS Num_Inmates_More_1_Year
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 365;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Num_Inmates_More_1_Year
0                     1355


### How Many Inmates have been in Custody for More Than a Month?

In [39]:
query = """SELECT COUNT(*) AS Num_Inmates_More_1_Month
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 30;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Num_Inmates_More_1_Month
0                      5392


### How Many Inmates have been in Custody for More Than 3 Months?

In [40]:
query = """SELECT COUNT(*) AS Num_Inmates_More_3_Months
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 90;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Num_Inmates_More_3_Months
0                       3936


### How Many Inmates have been in Custody for More Than 6 Months?

In [41]:
query = """SELECT COUNT(*) AS Num_Inmates_More_6_Months
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 180;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Num_Inmates_More_6_Months
0                       2691


### How Many Inmates have been in Custody for More Than 9 Months?

In [42]:
query = """SELECT COUNT(*) AS Num_Inmates_More_9_Months
            FROM mytable
            WHERE julianday('now') - julianday(Date) > 270;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

   Num_Inmates_More_9_Months
0                       1907


## Group 6: Miscellaneous 

### How Many Inmates Under Mental Observation are there for Each Custody Level?

In [38]:
query = """SELECT BRADH, CUSTODY_LEVEL,
            COUNT(*) AS Number_of_Inmates,
            COUNT(*) * 100.0 / (SELECT COUNT(*) FROM mytable) AS Percentage
            FROM mytable
            WHERE CUSTODY_LEVEL != 'None'
            GROUP BY CUSTODY_LEVEL, BRADH
            ORDER BY COUNT(*) DESC;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  BRADH CUSTODY_LEVEL  Number_of_Inmates  Percentage
0     Y           MED               1312   21.347218
1     N           MED               1302   21.184510
2     Y           MAX               1108   18.027986
3     N           MIN                920   14.969086
4     Y           MIN                806   13.114221
5     N           MAX                665   10.820046


For the most part, the number of individuals under mental health observation is fairly similar in each custody level except for maximum security. There is almost 2x more inmates under mental health observation in maximum security than those who aren't under supervision. 

### What are the Top 10 Charges for the Inmates Under Mental Health Observation?

In [52]:
query = """SELECT BRADH, TOP_CHARGE, COUNT(*) AS Count
            FROM mytable
            WHERE BRADH = 'Y' AND TOP_CHARGE != 'None'
            GROUP BY BRADH, TOP_CHARGE
            ORDER BY COUNT(*) DESC
            LIMIT 10;
        """
        
result = pd.read_sql_query(query, conn)
print(result)

  BRADH  TOP_CHARGE  Count
0     Y      125.25    539
1     Y  110-125.25    298
2     Y      120.05    255
3     Y      160.15    255
4     Y      265.03    160
5     Y      140.25    149
6     Y      160.10    129
7     Y      140.20    104
8     Y      120.10     91
9     Y      220.16     58


Here's a breakdown of the definition of each top charge according to [NYSenate.gov](https://www.nysenate.gov/legislation/laws/search/125.25):
* 125.25 - Murder in the Second Degree
* 110-125.25 - Attempted Murder in the Second Degree
* 120.05 - Assault in the Second Degree
* 160.15 - Robbery in the First Degree
* 265.03 - Criminal Possession of a Weapon in the Second Degree
* 140.25 - Burglary in the Second Degree
* 160.10 - Robbery in the Second Degree
* 140.20 - Burglary in the Third Degree
* 120.10 - Assault in the First Degree
* 220.16 - Criminal Possession of a Controlled Substance in the Third Degree

### Gang Affiliations by Gender

In [10]:
query = """SELECT GENDER, SRG_FLG, COUNT(*) AS Charge_Count
            FROM mytable
            WHERE GENDER IS NOT NULL
            GROUP BY GENDER, SRG_FLG
        """

result = pd.read_sql_query(query, conn)
print(result)

# calculations to figure out the % of each gender that are in a gang
female_gang_percentage = 23/382 * 100
male_gang_percentage = (1234/ (1234+4512)) * 100
print('\n')
print('Female Gang Member Percentage', female_gang_percentage)
print('Male Gang Member Percentage', male_gang_percentage)

  GENDER SRG_FLG  Charge_Count
0      F       N           359
1      F       Y            23
2      M       N          4512
3      M       Y          1234


Female Gang Member Percentage 6.020942408376963
Male Gang Member Percentage 21.47580925861469


A much higher percentage of male inmates have a gang affiliation compared to the female inmates.

### Inmates Under Mental Health Observation by Race

In [54]:
query = """SELECT RACE, BRADH, COUNT(*) AS Charge_Count 
            FROM mytable
            WHERE RACE IS NOT NULL
            GROUP BY RACE, BRADH
            ORDER BY Charge_Count DESC;
        """

result = pd.read_sql_query(query, conn) 
print(result)

   RACE BRADH  Charge_Count
0     B     Y          1961
1     B     N          1717
2     O     N           882
3     O     Y           881
4     W     Y           289
5     W     N           224
6     A     Y            77
7     A     N            69
8     I     Y            12
9     I     N             7
10    U     N             5
11    U     Y             4


For every race, the percentage of inmates who are in mental health observations is roughly around half. There doesn't appear to be any race that is over-represented among the inmates under mental health observation.

### Mental Health Status by Gang Affiliation

In [65]:
query = """SELECT SRG_FLG, BRADH, COUNT(*) AS Charge_Count
            FROM mytable
            GROUP BY SRG_FLG, BRADH
            ORDER BY Charge_Count DESC;
        """

result = pd.read_sql_query(query, conn) 
print(result)

gang_mental_obs = (678/(678+582)) * 100
non_gang_mental_obs = (2555/(2555+2331)) * 100

print('\n')
print('Non-Gang Affiliated Mental Health Percentage',non_gang_mental_obs)
print('Gang Affiliated Mental Health Percentage',gang_mental_obs)

  SRG_FLG BRADH  Charge_Count
0       N     Y          2555
1       N     N          2331
2       Y     Y           678
3       Y     N           582


Non-Gang Affiliated Mental Health Percentage 52.29226361031518
Gang Affiliated Mental Health Percentage 53.80952380952381


Gang affiliated inmates have a slightly higher occurrence of being under mental health observation. I was expecting a bigger disparity between the two.